### ISSUE TITLE
do not add final to abstract case classes

### ISSUE THREAD

**Issue by fommil** [scalacenter/scalafix#678]
related to #603

we have a pattern in scalaz to avoid an object alloc

```
  private sealed abstract case class Later[A]()             extends AsyncReturn[A]
  private object Later {
    // avoids object alloc, whilst allowing exhaustivity checking
    private[this] val value: Later[Nothing] = new Later[Nothing] {}
    def apply[A](): AsyncReturn[A]          = value.asInstanceOf[AsyncReturn[A]]
  }
```

but scalafix keeps wanting to add final.

**Comment by fommil** [comment]
furthermore, adding

```scala
// scalafix:off
...
// scalafix:on
```

or `// scalafix:ok`

is failing to disable the rewrite rule.

**Comment by fommil** [comment]
```@SuppressWarnings(Array("scalafix:MissingFinal"))``` also fails to avoid the rewrite.

**Comment by fommil** [comment]
I had to disable the rule in https://github.com/scalaz/effect/pull/9

**Comment by marcelocenerine** [comment]
From what I could see, the problem is not caused by a linter but a rewrite:
```
-  sealed abstract case class Later[A] private ()     extends AsyncReturn[A]
+  final sealed abstract case class Later[A] private ()     extends AsyncReturn[A]
```
Not sure how the suppression mechanism is supposed to work for rewrites (@olafurpg ?). After debugging for a little while I found out that the reason for the rule not being suppressed is that [this code](https://github.com/scalacenter/scalafix/blob/master/scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala#L48) only runs the suppressing logic if the patch being applied is an `AtomicPatch` or a `LintPatch`, which is not the case for your example (`TokenPatch.Add`). Olafur and @MasseGuillaume maybe know more details.

btw: support for `@SuppressWarnings` is available from [0.6.0-M2](https://github.com/scalacenter/scalafix/releases/tag/v0.6.0-M2)

**Comment by marcelocenerine** [comment]
ok, after thinking more about this and a good night of sleep:
- the suppression would work if the rule made the patch atomic
- indeed the the rule apparently lacks a check to skip the rewrite for `sealed abstract` classes inheriting from another `sealed trait` or `sealed abstract class`

I can create a PR with a fix later.

**Comment by fommil** [comment]
awesome, I'm really looking forward to that! #603 might be possible at the same time: you can't have `final` on an inner class.

**Comment by marcelocenerine** [comment]
Sure, I can look into both. Sorry for my comment last night. I completely overlooked your comments and was looking at the wrong place as I had many tickets related to the suppression mechanism simultaneously opened.

**Comment by fommil** [comment]
no need to apologise for anything! :smile:

**Comment by marcelocenerine** [comment]
I just created a PR (#706) fixing the 2 issues found here and in #603.

**Comment by olafurpg** [comment]
Fixed in #706
